### PR TITLE
prep extracellular potentials tutorial. Updated corresponding example

### LINKED ISSUE #1822: Add tutorial for extracellular signal predictions
## Goal
@brenthuisman suggested to promote https://github.com/arbor-sim/arbor/blob/master/python/example/single_cell_extracellular_potentials.py to a tutorial in this discussion: https://github.com/arbor-sim/arbor/discussions/1806

### PR DESCRIPTION
**Pull request by espenhgn**
Closes #1822. Adds a separate tutorial for extracellular signal predictions using LFPykit to the docs. Adds some minor improvements to the corresponding python example file.

### CODE REVIEW

File: doc/tutorial/tutorial_lfpykit.rst [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+.. literalinclude:: ../../python/example/single_cell_extracellular_potentials.py
+   :language: python
+   :lines: 24-56
+
+
+Load morphology on ``SWC`` file format:
+
+.. code-block:: python
+
+   # define morphology needed for ``arbor.place_pwlin`` and ``arbor.cable_cell`` below
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:58]: In other examples, the cell related stuff is generated from inside a function, providing some clarify in the code. I would argue to do that here too. It's lines 58-107; the function could return `p,cell`.

Also, is there anything to describe the cell textually? For instance, you add the `iclamp` in a specific place, I think this is quite tied to `single_cell_detailed.swc`. If so, then hardcode the filename + consider adding an illustration of that location (see `doc/concepts/labels.rst`, I can do it if needed.)

File: python/example/single_cell_extracellular_potentials.py [hunk 2]
```diff
@@ -122,100 +117,164 @@ def probes(self, gid):
...
+# extract time, V_m, I_m and I_c for each CV
 V_m_samples, V_m_meta = sim.samples(v_handle)[0]
 I_m_samples, I_m_meta = sim.samples(i_handle)[0]
 I_c_samples, I_c_meta = sim.samples(c_handle)[0]
 
-
 # drop recorded V_m values and corresponding meta data of
 # zero-sized CVs (branch-point potentials)
 inds = np.array([m.dist != m.prox for m in V_m_meta])
 V_m_samples = V_m_samples[:, np.r_[True, inds]]
```
Comment (brenthuisman) [inline python/example/single_cell_extracellular_potentials.py:137]: This line I do not understand, and the description on lines 134-135 only cover line 136, IIUC. Can you give a bit more detail?

File: python/example/single_cell_extracellular_potentials.py [hunk 3]
```diff
@@ -122,100 +117,164 @@ def probes(self, gid):
...
-
 # note: the cables comprising the metadata for each probe
 # should be the same, as well as the reported sample times.
 assert V_m_meta == I_m_meta
 assert (V_m_samples[:, 0] == I_m_samples[:, 0]).all()
 
-# prep recorded data for plotting
+# prep recorded data for plotting and computation of extracellular potentials
 time = V_m_samples[:, 0]
 V_m = V_m_samples[:, 1:].T
```
Comment (brenthuisman) [inline python/example/single_cell_extracellular_potentials.py:147]: Why is the transpose necessary?

File: python/example/single_cell_extracellular_potentials.py [hunk 4]
```diff
@@ -122,100 +117,164 @@ def probes(self, gid):
...
-dz = 1
-axis = np.round([x.min() - 10, x.max() + 10, y.min() - 10, y.max() + 10])
-# axis = np.round(axis)
-X, Y = np.meshgrid(np.linspace(axis[0], axis[1], int(np.diff(axis[:2]) // dx) + 1),
-                   np.linspace(axis[2], axis[3], int(np.diff(axis[2:]) // dz) + 1))
-Z = np.zeros_like(X)
+# ## Compute extracellular potentials
+# First we define a couple of classes to interface the LFPykit
+# library (https://LFPykit.readthedocs.io, https://github.com/LFPy/LFPykit):
+class ArborCellGeometry(lfpykit.CellGeometry):
```
Comment (brenthuisman) [inline python/example/single_cell_extracellular_potentials.py:161]: Nice! `ArborCellGeometry` and `ArborLineSourcePotential` seem like snippets that might eventually live inside a utility grab bag in either project. Not that it is important to move them there now, but what are your thoughts on that?

File: doc/tutorial/tutorial_lfpykit.rst [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+(`lfpykit.LineSourcePotential <https://lfpykit.readthedocs.io/en/latest/#class-linesourcepotential>`_).
+Other use cases may inherit from any other parent class defined in :class:`lfpykit.models`
+(`lfpykit.models <https://lfpykit.readthedocs.io/en/latest/#module-lfpykit.models>`_) in a similar manner:
+
+.. literalinclude:: ../../python/example/single_cell_extracellular_potentials.py
+   :language: python
+   :lines: 197-244
+
+
+With these two classes one may then compute extracellular potentials from transmembrane
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:121]:
```suggestion
With these two classes we can compute extracellular potentials from transmembrane
```

File: doc/tutorial/tutorial_lfpykit.rst [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+   :language: python
+   :lines: 247-270
+
+
+.. _tutorial_lfpykit-illustration:
+
+The result
+----------
+
+The visualization below of simulation results shows both the cellular geometry and a contour plot
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:134]:
```suggestion
The visualization below of simulation results shows the cellular geometry and a contour plot
```

File: doc/tutorial/tutorial_lfpykit.rst [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+
+.. Note::
+
+    The spatial discretization is here deliberately coarse with only 3 CVs per branch.
+    Hence the branch receiving input about 1/6 of the way from its root
+    (from ``decor.place('(location 4 0.16667)', iclamp, '"iclamp"')``) is treated as 3 separate
+    line sources with inhomogeneous current density per length unit. This inhomogeneity
+    is due to the fact that the total transmembrane current per CV may
+    distributed across multiple segments with varying surface area. The transmembrane
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:151]:
```suggestion
    be distributed across multiple segments with varying surface area. The transmembrane
```

File: doc/tutorial/tutorial_lfpykit.rst [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    current is assumed to be constant per length unit per segment.
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:152]:
```suggestion
    current is assumed to be constant per unit length per segment.
```
I think?

File: doc/tutorial/tutorial_lfpykit.rst [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    :width: 1600
+    :align: center
+
+
+.. Note::
+
+    The spatial discretization is here deliberately coarse with only 3 CVs per branch.
+    Hence the branch receiving input about 1/6 of the way from its root
+    (from ``decor.place('(location 4 0.16667)', iclamp, '"iclamp"')``) is treated as 3 separate
+    line sources with inhomogeneous current density per length unit. This inhomogeneity
```
Comment (brenthuisman) [inline doc/tutorial/tutorial_lfpykit.rst:149]:
```suggestion
    line sources with inhomogeneous current density per unit lenth. This inhomogeneity
```
I think?

File: doc/tutorial/index.rst [hunk 10]
```diff
@@ -7,7 +7,7 @@ Tutorials
     You can find some examples of full Arbor simulations in the ``python/examples`` directory of the
     `Arbor repository <https://github.com/arbor-sim/arbor>`_.
 
-    The examples use ``pandas`` and ``seaborn`` for analysis and plotting which are expected to be
+    The examples use ``pandas``, ``seaborn`` and ``LFPykit`` for analysis and plotting which are expected to be
```
Comment (brenthuisman) [inline doc/tutorial/index.rst:10]: `seaborn` and `pandas` are actually not imported in the script. I suppose the plot is not possible/hard to make with Seaborn primitives?

**Review by brenthuisman — CHANGES_REQUESTED** [review]
Thanks! Some feedback:

File: doc/tutorial/tutorial_lfpykit.rst [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+This example takes different elements from the above example tutorials
+(:ref:`A simple single cell model <tutorialsinglecell>`,
+:ref:`A detailed single cell recipe <tutorialsinglecellswcrecipe>`)
+to create a geometrically
+complex single cell model from an SWC morphology file, and adds predictions of
+extracellular potentials using the external `LFPykit <https://lfpykit.readthedocs.io/en/latest>`_ Python library
+(https://LFPykit.readthedocs.io, https://github.com/LFPy/LFPykit).
+`LFPykit <https://lfpykit.readthedocs.io/en/latest>`_ provides a few different classes facilitating
+calculations of extracellular potentials and related electroencephalography (EEG)
+and magnetoencephalography (MEG) signals from geometrically detailed neuron models under various assumptions.
```
[the remark this answers is not in the document]
Comment (espenhgn) [reply doc/tutorial/tutorial_lfpykit.rst:15]: Sure can do!

File: doc/tutorial/tutorial_lfpykit.rst [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+Define ``Recipe`` class:
+
+.. literalinclude:: ../../python/example/single_cell_extracellular_potentials.py
+   :language: python
+   :lines: 24-56
+
+
+Load morphology on ``SWC`` file format:
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (espenhgn) [reply doc/tutorial/tutorial_lfpykit.rst:55]: Ok.

File: doc/tutorial/index.rst (hunk 10 above)
Comment (espenhgn) [reply doc/tutorial/index.rst:10]: True, I don't use seaborn or pandas in the corresponding example, but the other examples do. One could set some `plt.rcParams` to make the plot more seaborn like I imagine (that is, if your main concern is that plots should have similar style across the documentation).

Comment (espenhgn) [reply doc/tutorial/index.rst:10]: I've updated the plot setting `plt.style.use('seaborn')`.

File: python/example/single_cell_extracellular_potentials.py (hunk 3 above)
Comment (espenhgn) [reply python/example/single_cell_extracellular_potentials.py:147]: I've removed the transpose for `V_m` and `I_m`. It was only there to make it's dimensionality match `I_m` which I also transposed to allow the multiplication with `M` below. I now transpose `I_m` in situ (so `V_e = M @ I_m.T`)

File: python/example/single_cell_extracellular_potentials.py (hunk 4 above)
Comment (espenhgn) [reply python/example/single_cell_extracellular_potentials.py:161]: Indeed. I think these can be put in LFPykit in due time. Perhaps there is a Python subclassing trick for all other classes that are set up similar to `LFPykit.LineSourcePotential` such that each modified class do not have to be explicitly defined. The modified method `get_transformation_matrix` would be the same in each case, as would be the `__init__` routine.

File: python/example/single_cell_extracellular_potentials.py (hunk 2 above)
Comment (espenhgn) [reply python/example/single_cell_extracellular_potentials.py:137]: I agree this was confusing (also momentarily confused me as the the voltage probe samples at branch points while the current probe does not). The `V_m' values are only used for the plot. I'll detail this some more.

File: python/example/single_cell_extracellular_potentials.py [hunk 13]
```diff
@@ -19,7 +19,6 @@
 import matplotlib.pyplot as plt
```
[the remark this answers is not in the document]
Comment (espenhgn) [reply python/example/single_cell_extracellular_potentials.py:18]: fixed.

File: doc/tutorial/tutorial_lfpykit.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (espenhgn) [reply doc/tutorial/tutorial_lfpykit.rst:15]: I've now included the entire line-source formalism. If this is too much detail, let me know.

File: doc/tutorial/tutorial_lfpykit.rst (hunk 1 above)
Comment (espenhgn) [reply doc/tutorial/tutorial_lfpykit.rst]: I've added a callable function `make_cable_cell` returning `p, cell` as suggested. 
If you see a neater way of setting a stimulus location as `clamp_location = arbor.location(4, 1/6)`, please revise. 
The location is now added to the results figure.

**Comment by espenhgn** [comment]
Hi again @brenthuisman . I think the points you raised should now be resolved.

**Review by brenthuisman — APPROVED** [review]
lgtm. Can you make one empty commit before I merge? Readthedocs should now be building for every PR.

**Comment by espenhgn** [comment]
> lgtm. Can you make one empty commit before I merge? Readthedocs should now be building for every PR.

Done!

**Comment by brenthuisman** [comment]
OK, there is a small config problem that prevents RTD from posting back the docs build result, but here it is: https://arbor--1825.org.readthedocs.build

I've spotted a few small things, I'm going to make suggestions for them.

**Pull request merged**